frame/im-online/src/lib.rs

```diff
@@ -260,7 +266,14 @@ pub mod pallet {
 			+ RuntimeAppPublic
 			+ Default
 			+ Ord
-			+ MaybeSerializeDeserialize;
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen;
+
+		/// The maximum number of keys that can be added.
+		type MaxKeys: Get<u32>;
+
+		/// The maximum number of peers to be stored in ReceivedHeartbeats
+		type MaxPeerInHeartbeats: Get<u32>;
```

Comment:
This doesn't describe its current usage. Currently this limit the number of bytes the encoding of the network state of the heartbeat.
Not the number of Peer.

Either we update the doc to say precisely what is this limit. But then this limit is a bit difficult to configure, or we update our usage of this limit.

We can do that by create a structure `BoundedHeartBeat<BlockNumber, Limit>` with the field `BoundedOpaqueNetworkState<Limit>` which would be generic over the limit. Maybe we can create a wrapper type to encode BoundedHeartBeat with the compact u32 in front of type. like a new struct `WrapperOpaque<Foo>(Foo)`.

So then we implement MaxEncodedLen on WrapperOpaque, BoundedHeartBeat and BoundedOpaqueNetworkState. And put into the storage the type `WrapperOpaque<BoundedHeartBeat<BlockNumber, MaxPeerInHearbeats>>`

(the type WrapperOpaque might be handy we can put it in frame_support probably)

Reply:
Good point, why do we need `BoundedHeartbeat` though?
Should we not be storing `WrapperOpaque<BoundedOpaqueNetworkState<_,_>>` instead?
Unless you are suggesting to change the signature of the extrinsic `heartbeat` but that is a bit of an overkill, no?

Reply:
Yes I mix myself, `WrapperOpaque<BoundedOpaqueNetworkState<_,_>>` is what we want indeed

Reply:
I am not too sure to understand what do you mean by WrapperOpaque? What is its purpose? Why not store directly `BoundedOpaqueNetwork` ?

Reply:
Yes WrapperOpaque was just a helper struct, in order not to change the encoding, (because currently it is encoded as a `Vec<u8>`).
I introduce this struct in this PR https://github.com/paritytech/substrate/pull/9738
So you could use `WrapperOpaque<BoundedNetworkState<..>>` where BoundedNetworkState is using a simple `#[derive(Encode,Decode)]` or `BoundedOpaqueNetworkState` but then you need to implement the encoding/decoding manually I think.

Reply:
Oh great, I will take a look and amend accordingly. It was a bit tricky this refactoring.
In particular, if I wanted to avoid adding some dependencies, PeerId should in fact have a bounded encoding, but it's difficult to add at source. MultiAddr could in theory be unbounded, but in practice would be, so I added one extra variable MaxPeerIdEncodingSize which represents the max encoding of these 2 variables. Please let me know if what I did is in the right track, and I will also refactor to include your inclusion of `WrapperOpaque`
Thanks!

Reply:
I have pushed through your addition of `WrapperOpaque` adding a few things to integrate it. Obviously, happy if you prefer to wait for your PR to go live first if you prefer.

Reply:
I think it is good to have it here so we can merge this PR first.